src/server/builder.ts

```diff
@@ -0,0 +1,417 @@
+                    array.splice(~insertIndex, 0, fileInfo);
+                }
+            }
+        }
+
+        static removeFromReferenceList(array: ModuleBuilderFileInfo[], fileInfo: ModuleBuilderFileInfo, afterFinalized?: boolean) {
+            if (!array) {
+                return;
+            }
+
+            if (!afterFinalized) {
+                // Before finalized the reference lists are not sorted
+                for (let i = 0; i < array.length; i++) {
+                    if (ModuleBuilderFileInfo.compareFileInfos(array[i], fileInfo) === 0) {
+                        array.splice(i, 1);
+                        break;
+                    }
+                }
+            }
+            else {
+                const removeIndex = binarySearch(array, fileInfo, ModuleBuilderFileInfo.compareFileInfos);
+                if (removeIndex >= 0) {
+                    array.splice(removeIndex, 1);
+                }
+            }
+        }
+
+        addReferencedBy(fileInfo: ModuleBuilderFileInfo): void {
+            if (this.referencedBy.length === 0) {
+                this.referencedBy.push(fileInfo);
+                return;
+            }
+            ModuleBuilderFileInfo.addToReferenceList(this.referencedBy, fileInfo, this.finalized);
+        }
+
+        removeReferencedBy(fileInfo: ModuleBuilderFileInfo): void {
+            ModuleBuilderFileInfo.removeFromReferenceList(this.referencedBy, fileInfo, this.finalized);
+        }
+
+        finalize() {
+            if (this.finalized) {
+                return;
+            }
+
+            if (this.references.length > 0) {
+                this.references.sort(ModuleBuilderFileInfo.compareFileInfos);
+            }
+            if (this.referencedBy.length > 0) {
+                this.referencedBy.sort(ModuleBuilderFileInfo.compareFileInfos);
+            }
+            this.finalized = true;
+        }
+    }
+
+    class ModuleBuilder extends AbstractBuilder<ModuleBuilderFileInfo> {
+
+        private isDependencyGraphBuilt = false;
+        private lastUpdateProjectVersion: string;
+
+        getReferencedFileInfos(fileName: string): ModuleBuilderFileInfo[] {
+            const referencedFileNames = this.project.getReferencedFiles(fileName);
+            if (referencedFileNames && referencedFileNames.length > 0) {
+                return map(referencedFileNames, f => this.getFileInfo(f)).sort(ModuleBuilderFileInfo.compareFileInfos);
+            }
+            return [];
+        }
+
+        ensureDependencyGraph() {
+            if (this.isDependencyGraphBuilt) {
+                return;
+            }
+
+            const fileNames = this.project.getFileNames();
+            for (const fileName of fileNames) {
+                const sourceFile = this.project.getSourceFile(fileName);
+                if (sourceFile) {
+                    const newFileInfo = new ModuleBuilderFileInfo(this.project.getScriptInfo(fileName), this.project);

```

Comment:
this code is similar to what you already have in `NonModuleBuilder.getOrCreateScriptInfo`. Consider moving this method to the base class and parameterize base class with constructor function to create `BuilderFileInfo` objects.